packages/edit-site/src/components/dataviews/provider.js

```diff
@@ -0,0 +1,253 @@
+			}
+			const dataViewQuery = {
+				wp_dataviews_type: typeTaxonomyId,
+				orderby: 'date',
+				order: 'asc',
+			};
+			return {
+				dataViewRecords: getEntityRecords(
+					'postType',
+					'wp_dataviews',
+					dataViewQuery
+				),
+				dataViewIsLoading: isResolving( 'getEntityRecords', [
+					'postType',
+					'wp_dataviews',
+					dataViewQuery,
+				] ),
+				dataViewIsSaving: isSavingEntityRecord(
+					'postType',
+					'wp_dataviews'
+				),
+			};
+		},
+		[ typeTaxonomyId ]
+	);
+	const { saveEntityRecord } = useDispatch( coreStore );
+	useEffect( () => {
+		if (
+			dataViewRecords?.length === 0 &&
+			! dataViewIsLoading &&
+			! dataViewIsSaving &&
+			! isCreatingADefaultView.current
+		) {
+			isCreatingADefaultView.current = true;
+			saveEntityRecord( 'postType', 'wp_dataviews', {
+				title: 'All',
+				status: 'publish',
+				wp_dataviews_type: typeTaxonomyId,
+				content: JSON.stringify( DEFAULT_VIEWS[ type ] ),
+			} );
+		}
+	}, [
+		type,
+		dataViewIsLoading,
+		dataViewRecords,
+		dataViewIsSaving,
+		typeTaxonomyId,
+		saveEntityRecord,
+	] );
+	useEffect( () => {
+		if ( dataViewRecords?.length > 0 ) {
+			isCreatingADefaultView.current = false;
+		}
+	}, [ dataViewRecords ] );
+	useEffect( () => {
+		isCreatingADefaultView.current = false;
+	}, [ typeTaxonomyId ] );
+	if ( dataViewRecords?.length > 0 ) {
+		return dataViewRecords;
+	}
+	return null;
+}
+
+function DataviewsProviderInner( { type, children } ) {
+	const [ currentViewId, setCurrentViewId ] = useState( null );
+	const dataviewTypeTaxonomyId = useDataviewTypeTaxonomyId( type );
+	const dataviews = useDataviews( type, dataviewTypeTaxonomyId );
+	const { editEntityRecord } = useDispatch( coreStore );
+	useEffect( () => {
+		if ( ! currentViewId && dataviews?.length > 0 ) {
+			setCurrentViewId( dataviews[ 0 ].id );
+		}
+	}, [ currentViewId, dataviews ] );
+	const editedViewRecord = useSelect(
+		( select ) => {
+			if ( ! currentViewId ) {
+				return;
+			}
+			const { getEditedEntityRecord } = select( coreStore );
+			const dataviewRecord = getEditedEntityRecord(
+				'postType',
+				'wp_dataviews',
+				currentViewId
+			);
+			return dataviewRecord;
+		},
+		[ currentViewId ]
+	);
+
+	const value = useMemo( () => {
+		return {
+			taxonomyId: dataviewTypeTaxonomyId,
+			dataviews,
+			currentViewId,
+			setCurrentViewId,
+			view: editedViewRecord?.content
+				? JSON.parse( editedViewRecord?.content )
+				: DEFAULT_VIEWS[ type ],
+			setView( view ) {
+				if ( ! currentViewId ) {
+					return;
+				}
+				editEntityRecord( 'postType', 'wp_dataviews', currentViewId, {
+					content: JSON.stringify( view ),
+				} );
+			},
+		};
+	}, [
+		type,
+		dataviewTypeTaxonomyId,
+		dataviews,
+		currentViewId,
+		editedViewRecord?.content,
+		editEntityRecord,
+	] );
+
+	return (
+		<DataviewsContext.Provider value={ value }>
+			{ children }
+		</DataviewsContext.Provider>
+	);
+}
+export default function DataviewsProvider( { children } ) {
```

Comment:
Am I correct in assuming that the decision to use a provider is that the context will be managed separately by the main view and the sidebar?


My first impressions: there is a lot of code in this file and I wonder if all the complexity is absolutely necessary.

Reply:
> Am I correct in assuming that the decision to use a provider is that the context will be managed separately by the main view and the sidebar?

@mcsf Exactly both the sidebar and the main view need to change the active view access the views etc so I needed a provider to share this information/state across different parts of the UI.

> there is a lot of code in this file and I wonder if all the complexity is absolutely necessary.

The code contains a lot of complexity, but it is necessary. For example, we need to create a taxonomy that can identify a view for a type such as a page. If the taxonomy does not exist, we need to create it. The logic behind this is that we need to wait for the request to determine if the taxonomy already exists. If it doesn't, we create it, and the same for a post with a default view. Although I considered using actions and resolvers to implement this, it did not seem simple, and I was not sure how to follow that approach without previous history. To make the logic easier to understand, I will add better comments, and I will attempt to reuse more logic between creating a taxonomy and creating a post.